docs: add upgrade notes for 0.9 - 0.10 (#5732)
confluentinc/ksql: CHANGELOG.md

* /inserts-stream endpoint now supports nested types ([#5621](https://github.com/confluentinc/ksql/pull/5621)) ([866ae34](https://github.com/confluentinc/ksql/commit/866ae3499a9d59068461d17ed15e1353759a0334))
* don't fail if broker does not support AuthorizedOperations ([#5617](https://github.com/confluentinc/ksql/pull/5617)) ([0feb081](https://github.com/confluentinc/ksql/commit/0feb081c89d00f326e151035ac5616d19f610307))
* ensure only deserializable cmds are written to command topic ([#5645](https://github.com/confluentinc/ksql/pull/5645)) ([4ad2bde](https://github.com/confluentinc/ksql/commit/4ad2bde44f31ea3c15fdd898ce12d50a595013d1))
* fail AVRO/Protobuf/JSON Schema statements if SR is missing ([#5597](https://github.com/confluentinc/ksql/pull/5597)) ([85a0320](https://github.com/confluentinc/ksql/commit/85a0320e774d236bda4898f82eb5d5c2f7b42cec))
* support GROUP BY with no source columns used ([#5644](https://github.com/confluentinc/ksql/pull/5644)) ([a8e6630](https://github.com/confluentinc/ksql/commit/a8e66304081879335c62948d5a8dd5f5531766be))

### BREAKING CHANGES

#### Any key name

> Statements containing PARTITION BY, GROUP BY or JOIN clauses will now produce different output schemas.

Reviewer: ```suggestion
Statements containing PARTITION BY, GROUP BY, or JOIN clauses now produce different output schemas.

```

For PARTITION BY and GROUP BY statements, the name of the key column in the result is determined by the PARTITION BY or GROUP BY clause:

> 1. Where the partitioning or grouping is a single column reference, then the key column will have the same name as this column. For example:

Reviewer: ```suggestion
1. Where the partitioning or grouping is a single column reference, then the key column has the same name as this column. For example:

```

```sql
CREATE STREAM OUTPUT AS
SELECT * FROM INPUT GROUP BY X;
-- OUTPUT will have a key column called X;
```

> 2. Where the partitioning or grouping is a single struct field, then the key column will have the same name as the field. For example:

Reviewer: ```suggestion
2. Where the partitioning or grouping is a single struct field, then the key column has the same name as the field. For example:

```

```sql
CREATE STREAM OUTPUT AS
SELECT * FROM INPUT GROUP BY X->field1;
-- OUTPUT will have a key column called FIELD1;
```

> 3. Otherwise, the key column name will be system generated and be in the form `KSQL_COL_n`, where `n` is some positive integer.

Reviewer: ```suggestion
3. Otherwise, the key column name is system-generated and has the form `KSQL_COL_n`, where `n` is a positive integer.

```

> In all case, except where grouping by more than one column, the new key column's name can be set by defining an alias in the projection. For example:

Reviewer: ```suggestion
In all cases, except where grouping by more than one column, you can set the new key column's name by defining an alias in the projection. For example:

```

```sql
CREATE TABLE OUTPUT AS
SELECT USERID AS ID, COUNT(*) FROM USERS GROUP BY ID;
-- OUTPUT will have a key column named ID.
```

> For groupings of multiple expressions there is currently no way to provide a name for the system generated key column. This is a shortcoming that will be fixed shortly when ksqlDB supports more than just a single key column.

Reviewer: ```suggestion
For groupings of multiple expressions, you can't provide a name for the system-generated key column.

```

Reviewer: In general, we don't to make promises about future features in docs.

For JOIN statements, the name of the key column in the result is determined by the join criteria.

> 1. For INNER and LEFT OUTER joins where the join criteria contains at least one column reference the key column will be named of the left most source whose join criteria is a column reference. For example:

Reviewer: ```suggestion
1. For INNER and LEFT OUTER joins where the join criteria contain at least one column reference, the key column is named based on the left-most source whose join criteria is a column reference. For example:

```

```sql
CREATE TABLE OUTPUT AS
SELECT * FROM I1 JON I2 ON abs(I1.ID) = I2.ID JOIN I3 ON I2.ID = I3.ID;
-- OUTPUT will have a key column named I2_ID.
```
The key column can be given a new name, if required, by defining an alias in the projection. For example:
```sql
CREATE TABLE OUTPUT AS
SELECT I2.ID AS ID, I1.V0, I2.V0, I3.V0 FROM I1 JON I2 ON abs(I1.ID) = I2.ID JOIN I3 ON I2.ID = I3.ID;
-- OUTPUT will have a key column named ID.
```

> 2. For FULL OUTER joins and other joins where the join criteria is not on column references, the key column in the output is not equivalent to any column from any source. The key column will have a system generated name in the form `KSQL_COL_n`, where `n` is a positive integer. For example:

Reviewer: ```suggestion
2. For FULL OUTER joins and other joins where the join criteria are not on column references, the key column in the output is not equivalent to any column from any source. The key column has a system-generated name in the form `KSQL_COL_n`, where `n` is a positive integer. For example:

```

```sql
CREATE TABLE OUTPUT AS
SELECT * FROM I1 FULL OUTER JOIN I2 ON I1.ID = I2.ID;
-- OUTPUT will have a key column named KSQL_COL_0, or similar.
```
The key column can be given a new name, if required, by defining an alias in the projection. A new UDF has been introduced to help define the alias called `JOINKEY`. It takes the join criteria as its parameters. For example:
```sql
CREATE TABLE OUTPUT AS
SELECT JOINKEY(I1.ID, I2.ID) AS ID, I1.V0, I2.V0 FROM  I1 FULL OUTER JOIN I2 ON I1.ID = I2.ID;
-- OUTPUT will have a key column named ID.
```
`JOINKEY` will be deprecated in a future release of ksqlDB once multiple key columns are supported.

#### Explicit keys

`CREATE TABLE` statements will now fail if the `PRIMARY KEY` column is not provided.

For example, a statement such as:

```sql
CREATE TABLE FOO (name STRING) WITH (kafka_topic='foo', value_format='json');
```

> Will need to be updated to include the definition of the PRIMARY KEY, e.g.

Reviewer: ```suggestion
Will need to be updated to include the definition of the PRIMARY KEY, for example:

```

```sql
CREATE TABLE FOO (ID STRING PRIMARY KEY, name STRING) WITH (kafka_topic='foo', value_format='json');
```

> If using schema inference, i.e. loading the value columns of the topic from the Schema Registry, the primary key can be provided as a partial schema, e.g.

Reviewer: ```suggestion
If using schema inference, i.e. loading the value columns of the topic from the Schema Registry, the primary key can be provided as a partial schema, for example:

```

```sql
-- FOO will have value columns loaded from the Schema Registry
CREATE TABLE FOO (ID INT PRIMARY KEY) WITH (kafka_topic='foo', value_format='avro');
```

> `CREATE STREAM` statements that do not define a `KEY` column will no longer have an implicit `ROWKEY` key column.

Reviewer: ```suggestion
`CREATE STREAM` statements that do not define a `KEY` column no longer have an implicit `ROWKEY` key column.

```

For example:

```sql
CREATE STREAM BAR (NAME STRING) WITH (...);
```

> The above statement would previously have resulted in a stream with two columns: `ROWKEY STRING KEY` and `NAME STRING`.

Reviewer: ```suggestion
Previously, the above statement would have resulted in a stream with two columns: `ROWKEY STRING KEY` and `NAME STRING`.

```

> With this change the above statement will result in a stream with only the `NAME STRING` column.

Reviewer: ```suggestion
With this change, the previous statement results in a stream with only the `NAME STRING` column.

```

> Streams will no KEY column will be serialized to Kafka topics with a `null` key.

Reviewer: ```suggestion
Streams with no KEY column are serialized to Kafka topics with a `null` key.

```

#### Key columns required in projection

A statement that creates a materialized view must include the key columns in the projection. For example:

```sql
CREATE TABLE OUTPUT AS
   SELECT 
      productId,  // <-- key column in projection
      SUM(quantity) as unitsSold
   FROM sales
   GROUP BY productId;
```

> The key column `productId` is required in the projection. In previous versions of ksqlDB the presence

Reviewer: ```suggestion
The key column `productId` is required in the projection. In previous versions of ksqlDB, the presence

```

of `productId` in the projection would have placed a _copy_ of the data into the value of the underlying 

> Kafka topic's record.  However, as of v0.10 the projection must include the key columns and these columns

Reviewer: ```suggestion
Kafka topic's record.  But starting in version 0.10.0, the projection must include the key columns, and these columns

```

> will be stored in the _key_ of the underlying Kafka record.  Optionally, you may provide an alias for 

Reviewer: ```suggestion
in the _key_ of the underlying Kafka record.  Optionally, you may provide an alias for 

```

the key column(s). 

```sql
CREATE TABLE OUTPUT AS
   SELECT 
      productId as id,  // <-- aliased key column
      SUM(quantity) as unitsSold
   FROM sales
   GROUP BY productId;
```

> If you require a copy of the key column in the Kafka record's value then you can use the 

Reviewer: ```suggestion
If you need a copy of the key column in the Kafka record's value, use the 

```

[AS_VALUE](docs/developer-guide/ksqldb-reference/scalar-functions#as_value) function to indicate this

> to ksqlDB. For example, the following will produce an output inline with the previous version of ksqlDB

Reviewer: ```suggestion
to ksqlDB. For example, the following statement produces an output inline with the previous version of ksqlDB

```

for the above example materialized view:

```sql
CREATE TABLE OUTPUT AS
   SELECT 
      productId as ROWKEY,              // <-- key column named ROWKEY
      AS_VALUE(productId) as productId, // <-- productId copied into value
      SUM(quantity) as unitsSold
   FROM sales
   GROUP BY productId;
```

### WITH(KEY) syntax removed

> In previous versions, all key columns where called `ROWKEY`. To enable a more

Reviewer: ```suggestion
In previous versions, all key columns were called `ROWKEY`. To enable a more

```

> user-friendly name to be used for the key column in queries it was possible

Reviewer: ```suggestion
user-friendly name for the key column in queries, it was possible

```

to supply an alias for the key column in the WITH clause, for example:

```sql
CREATE TABLE INPUT (ROWKEY INT PRIMARY KEY, ID INT, V0 STRING) WITH (key='ID', ...);
```

With the previous query, the `ID` column can be used as an alias for `ROWKEY`.

> This approach required the {{ site.ak }} message value to contain an exact copy

Reviewer: ```suggestion
This approach required the Kafka message value to contain an exact copy

```

Reviewer: This file isn't part of the doc build, so sitevars won't work.

of the key.

[KLIP-24](https://github.com/confluentinc/ksql/blob/master/design-proposals/klip-24-key-column-semantics-in-queries.md)
removed the restriction that key columns must be named `ROWKEY`, negating the need for the `WITH(KEY)`
syntax, which has been removed. Also, this change removed the requirement for

> the {{ site.ak }} message value to contain an exact copy of the key.

Reviewer: ```suggestion
the Kafka message value to contain an exact copy of the key.

```

> Update your queries by removing the `KEY` from the `WITH` clause and  naming

Reviewer: ```suggestion
Update your queries by removing the `KEY` from the `WITH` clause and naming

```

your `KEY` and `PRIMARY KEY` columns appropriately. For example, the previous
CREATE TABLE statement can now be rewritten as:

```sql
CREATE TABLE INPUT (ID INT PRIMARY KEY, V0 STRING) WITH (...);
```

Unless the value format is `DELIMITED`, which means the value columns are

> *order dependant*, so dropping the `ID` value column would result in a

Reviewer: ```suggestion
*order dependent*, so dropping the `ID` value column would result in a

```
